src/axom/core/Array.hpp

```diff
@@ -1006,37 +1133,42 @@ inline typename Array<T>::ArrayIterator Array<T>::erase(
 }
 
 //------------------------------------------------------------------------------
-template <typename T>
+template <typename T, IndexType dim>
 template <typename... Args>
-inline void Array<T>::emplace(IndexType pos, Args&&... args)
+inline void Array<T, dim>::emplace(IndexType pos, Args&&... args)
 {
   reserveForInsert(1, pos);
   m_data[pos] = std::move(T(std::forward<Args>(args)...));
 }
 
 //------------------------------------------------------------------------------
-template <typename T>
+template <typename T, IndexType dim>
 template <typename... Args>
-inline typename Array<T>::ArrayIterator Array<T>::emplace(Array<T>::ArrayIterator pos,
-                                                          Args&&... args)
+inline typename Array<T, dim>::ArrayIterator Array<T, dim>::emplace(
+  Array<T, dim>::ArrayIterator pos,
+  Args&&... args)
 {
   assert(pos >= begin() && pos <= end());
   emplace(pos - begin(), args...);
   return pos;
 }
 
 //------------------------------------------------------------------------------
-template <typename T>
+template <typename T, typename ArrayType>
 template <typename... Args>
-inline void Array<T>::emplace_back(Args&&... args)
+inline void Only1D<T, 1, ArrayType>::emplace_back(Args&&... args)
 {
-  emplace(m_num_elements, args...);
+  asDerived().emplace(asDerived().fullSize(), args...);
 }
 
 //------------------------------------------------------------------------------
-template <typename T>
-inline void Array<T>::resize(IndexType new_num_elements)
+template <typename T, IndexType dim>
+template <typename... Args>
+inline void Array<T, dim>::resize(Args... args)
 {
+  static_assert(sizeof...(Args) == dim,
+                "Array size must match number of dimensions");
+  const auto new_num_elements = detail::packProduct(args...);
   assert(new_num_elements >= 0);
```

Comment:
small nit pick: this assertion will pass with negative sizes as long as there's an even number of them.
```c++
assert(((args > 0) && ...));
```

Reply:
This is a little tricky to do in C++11 but I will put something in